File: setup_file.py
"""Driver code for automated installation.

This program will install Talawa Admin on the system.
It executes the pre-defined commands on the shell and
then logs the output of each step.The application runs
on a virtual environment created using Python. Every
installation process shows a message - whether the
installation was successful or not.

"""

import os
import asyncio
import pathlib
import platform
from setup import utils, installation

# 1. Shift to talawa-admin Directory
path = pathlib.Path(__file__).parent.resolve()
os.chdir(path)

# 2. Display initial data
utils.display_markdown("# TALAWA-ADMIN")
with open("./setup/markdown/About.md", encoding="utf-8") as about:
    utils.display_markdown(about.read(), "white")

# 3. Installing a virtual enviroment
utils.display_markdown("# SETTING UP ENVIROMENT")
VIRTUAL_ENV_INSTALLATION_COMMAND = 'pip install virtualenv && python -m pip install -U pip'
if platform.system() == "Debian":
    VIRTUAL_ENV_INSTALLATION_COMMAND = 'sudo apt install virtualenv'
elif platform.system() == "Linux":
    VIRTUAL_ENV_INSTALLATION_COMMAND = 'sudo apt install python3-virtualenv'
asyncio.run(
    installation.run(
        VIRTUAL_ENV_INSTALLATION_COMMAND,
        "Successfully installed [bold]virtualenv[/bold] :party_popper:",
        "Failed to create a virtual environment :cross_mark:"
    )
)

# 4. Creating a virtual enviroment
VIRTUAL_ENV_CREATION_COMMAND = 'virtualenv venv'
if platform.system() != "Windows":
    VIRTUAL_ENV_CREATION_COMMAND = 'virtualenv venv -p python3'

asyncio.run(
    installation.run(
        VIRTUAL_ENV_CREATION_COMMAND,
        "Successfully created a [bold]virtual environment[/bold] :party_popper:",
        "Failed to create a virtual environment :cross_mark:"
    )
)

# 5. Activating virtual environment
VIRTUAL_ENV_PATH = '.\\venv\\Scripts\\activate'
if platform.system() != "Windows":
    VIRTUAL_ENV_PATH = '. venv/bin/activate'

asyncio.run(
    installation.run(
        VIRTUAL_ENV_PATH,
        "Successfully activated [bold]virtual environment[/bold] :party_popper:",
        "Failed to activate virtual environment :cross_mark:"
    )
)

# 6. Installation of Python packages
utils.display_markdown("# INSTALLATION OF PYTHON PACKAGES")

File: setup_file.js
/**
 * @brief Driver code for express installation
 * @description This code installs the Talawa API by
 * running the individual steps automatically. The
 * process consists of 4 steps:
 * 1. Displaying information about Talawa API
 * 2. Install project dependencies
 * 3. Set up user configuration
 * 4. Display command to start the application
 *
 * Every step displays whether it has been
 * executed successfully or not
 */
import path from 'path';
import display_about from './setup/about.js';
import install_dependencies from './setup/package_installation.js';
import user_input from './setup/input.js';
import start_application from './setup/start_app.js';

/**
 * This asynchronous function runs the setup process
 * by executing each of the steps serially
 */
const run_setup = async () => {
  //1. Display information about the project
  display_about();

  //2. Install project dependencies
  await install_dependencies();

  //3. Set up user configuration
  const __dirname = path.resolve();
  await user_input(path.join(__dirname, '.env'));

  //4. Display command to start the application
  await start_application();
};

run_setup();
